text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+additional support specifically for crABI, to make crABI even more usable.
+
+In particular, some aspects of crABI define translations of types that involve
+generics or the generation of various small one-off data structures, and a
+language or tool with native support could handle these more conveniently
+without requiring the explicit separate definition of as many structures.
+
+Furthermore, native support for crABI could distinguish between owned and
+borrowed types, and handle those automatically in a way that makes sense for
+the language or tool. (For instance, native support could adopt owned types
+into whatever mechanism the language or tool uses to free objects when done
+when them.)
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Guaranteed niche optmization is the most uncertain part of the proposed crABI
+specification. crABI either needs to be layout-compatible with Rust `Option`
+and `Result` and tuple types, or needs to use its own distinct layout and thus
+distinct type for them, in which case they require conversion to/from the Rust
+types. However, Rust does not provide stability guarantees for its niche
+optimizations, so crABI cannot simultaneously have a documented ABI (defining
+an exact set of specific niche optimizations and no others) and maintain layout
+compatibility with Rust: Rust could not then do any further niche optimizations
+not already specified by crABI, even in an edition (because types and their
+layouts are intercompatible between editions). Thus, this crABI proposal
+(reluctantly) specifies *separate* `Option`, `Result`, and tuple types, to
+avoid this compatibility hazard.
+
+Making a structure compatible with crABI requires declaring it with either
+`repr(crabi)` or `repr(C)`. This makes it difficult to adapt a type from
+another library, or to add a crABI interface to an existing Rust library while
+preserving its data types.
+
+crABI could (instead of or in addition to `repr(crabi)`) provide a mechanism
+that "transforms" a type into the `repr(crabi)` equivalent of that type, to avoid
+having to define a separate type. However, this would not handle the problem of
+translating between the types, which in the general case of a "deep"
+translation could require copying arbitrary in-memory data structures. crABI
+does not aim to solve that problem.
+
+The translation of slices and similar uses structs containing pointer/length
+pairs, rather than inlining the pointer and length as separate arguments.
+[As noted above][types], this is typically passed and returned in an efficient
+fashion on major targets. However, in some languages, such as C, this will
+require separately defining a structure and then using that structure. This
+still seems preferable, though, as combining the two into one struct allows for
+uniform handling between arguments, return values, and fields, as well as
+keeping the pointer and length more strongly associated.
```

Comment:
Passing slices as two arguments (`ptr` then `len`) would allow Rust bindings to existing C code that does this to use slices directly. For example, POSIX `ssize_t send(int sockfd, const void buf, size_t len, int flags)` could be exposed in Rust as

```rust
extern "crABI" { // or perhaps even "C" ?
    fn send(socket: c_int, buf: *const [u8], flags: c_int) -> ssize_t;
}
```

Reply:
passing a structure of two fields is generally different than passing two arguments though, no?

Reply:
Yes, it is generally different. My suggestion is to make them the same for slices with `extern "crABI"`/`extern "C"`, as a special case only for that specific combination.